File: src/main/java/org/radarbase/management/service/mapper/decorator/UserMapperDecorator.java
package org.radarbase.management.service.mapper.decorator;

import org.radarbase.management.domain.User;
import org.radarbase.management.domain.audit.EntityAuditInfo;
import org.radarbase.management.service.RevisionService;
import org.radarbase.management.service.dto.UserDTO;
import org.radarbase.management.service.mapper.UserMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;

public abstract class UserMapperDecorator implements UserMapper {

    @Autowired
    @Qualifier("delegate")
    private UserMapper delegate;

    @Autowired
    private RevisionService revisionService;

    @Override
    public UserDTO userToUserDTO(User user) {
        if (user == null) {
            return null;
        }

        UserDTO dto = delegate.userToUserDTO(user);

        EntityAuditInfo auditInfo = revisionService.getAuditInfo(user);
        dto.setCreatedDate(auditInfo.getCreatedAt());
        dto.setCreatedBy(auditInfo.getCreatedBy());
        dto.setLastModifiedDate(auditInfo.getLastModifiedAt());
        dto.setLastModifiedBy(auditInfo.getLastModifiedBy());

        return dto;
    }
}

File: src/main/java/org/radarbase/management/service/mapper/decorator/UserMapperDecorator.kt
package org.radarbase.management.service.mapper.decorator

import org.radarbase.management.domain.User
import org.radarbase.management.service.RevisionService
import org.radarbase.management.service.dto.UserDTO
import org.radarbase.management.service.mapper.UserMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier

abstract class UserMapperDecorator : UserMapper {
    @Autowired
    @Qualifier("delegate")
    private lateinit var delegate: UserMapper

    @Autowired
    private lateinit var revisionService: RevisionService

    override fun userToUserDTO(user: User?): UserDTO? {
        if (user == null) {
            return null
        }
        val dto = delegate.userToUserDTO(user)
        val auditInfo = revisionService.getAuditInfo(user)
        dto?.createdDate = auditInfo.createdAt
        dto?.createdBy = auditInfo.createdBy
        dto?.lastModifiedDate = auditInfo.lastModifiedAt
        dto?.lastModifiedBy = auditInfo.lastModifiedBy
        return dto
    }
}
